README.md
# Refresher

My custom Pull to Refresh Widget, mimicking iOS Pull to Refresh except for the toolbar

## Installation

First, add `refresher` as a [dependency in your pubspec.yaml file](https://flutter.io/platform-plugins/).

```
refresher: ^0.0.4+2
```

## Example
```
class RefresherDemo extends StatefulWidget {
  @override
  _RefresherDemoState createState() => _RefresherDemoState();
}

class _RefresherDemoState extends State<RefresherDemo> with SingleTickerProviderStateMixin {
  String _title = "Refresher Demo";

  double percentage = 0.0;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text(_title), actions: [
        IconButton(
          icon: Icon(Icons.navigate_next),
          onPressed: () {
            Navigator.push(context,
                MaterialPageRoute(builder: (BuildContext context) => InfiniteRefresherDemo()));
          },
        )
      ]),
      body: Refresher(
          loadingSize: 50.0,
          margin: EdgeInsets.all(8.0),
          onRefresh: () async {
            int counter = 0;
            Timer.periodic(Duration(milliseconds: 1000), (t) {
              counter++;
              if (this.mounted)
                setState(() {
                  _title = "Fetching${List.filled(counter, ".").join()}";
                  if (counter == 5) t.cancel();
                });
            });
            await Future.delayed(new Duration(milliseconds: 6000));
            if (this.mounted)
              setState(() {
                _title = "Refresher Demo";
              });
          },
          child: Column(
              children: List.generate(
            100,
            (index) => Text("Here is $index"),
          ))),
    );
  }
}
```

## Infinite Example
```
class InfiniteRefresherDemo extends StatefulWidget {
  @override
  _InfiniteRefresherDemoState createState() => _InfiniteRefresherDemoState();
}

class _InfiniteRefresherDemoState extends State<InfiniteRefresherDemo>
    with SingleTickerProviderStateMixin {
  String _title = "Infinite Refresher Demo";
  List<Widget> _children = [];
  double percentage = 0.0;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(_title),
      ),
      body: InfiniteRefresher(
        loadingSize: 50.0,
        margin: EdgeInsets.all(8.0),
        onRefresh: () async {
          int counter = 0;
          Timer.periodic(Duration(milliseconds: 1000), (t) {
            counter++;
            if (this.mounted)
              setState(() {
                _title = "Fetching${List.filled(counter, ".").join()}";
                if (counter == 5) t.cancel();
              });
          });
          await Future.delayed(new Duration(milliseconds: 1000));
          _children.clear();
          _children.addAll(List.generate(
            Random().nextInt(100),
                (index) => Text("Here is $index"),
          ));
          if (this.mounted)
            setState(() {
              _title = "Infinite Refresher Demo";
            });
        },
        widgetBuilder: (context) {
          return Column(children: _children);
        },
        fetcher: (context, cursor) async {
          await Future.delayed(Duration(milliseconds: 3000));

          _children.addAll(List.generate(
            Random().nextInt(100),
            (index) => Text("Here is $index"),
          ));

          return true;
        },
      ),
    );
  }
}
```


CHANGELOG.md
## 1.0.1+1 - 19 februari 2020

* merge all typedefs

## 1.0.1 - 15 februari 2020

* add refresher builder

## 1.0.0+2

* Add LoadingAnimation.defaultLineColor and LoadingAnimation.defaultProgressColor

## 1.0.0+1

* Fix scroll to the max
* Fix refresher

## 1.0.0

* Stable version
* Add cursor and noMoreData arguments on InfiniteListView and RefresherInfinite to resume progress whenever changing tab
* Add flag should scroll to the max whenever scrolling to bottom most

## 0.0.4+2

* Refresher and Infinite Refresher only supports vertical axis
* Because NotificationListener<ScrollNotification> will be triggered too when its child/ren scroll
* Ideally there will be no vertical scrollable inside another vertical scrollable
* that's why to prevent this issue, _handleScrollNotification only fired when it's vertical scrolling

## 0.0.4+1

* Documentation fix

## 0.0.4

* Fix Android AlwaysScrollPhysics clamping problem.

## 0.0.3+2

* Disable Refresh when doing Fetch More Data and vice versa.

## 0.0.3+1

* Add new reset function without fetch new data (because it's handled at the front).

## 0.0.3

* Add Refresher combined with InfiniteListView (InfiniteRefresher).

## 0.0.2

* Add margin and loading size to argument, also scrollController is injectable from the outside world.
* Temporarily scroll functionality is disabled after it reaches for refresh height.

## 0.0.1+1

* Minor bug fix.

## 0.0.1

* Initial release.


example/lib/main.dart
import 'dart:async';
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:refresher/refresher.dart';
import 'package:refresher_example/infinite_demo.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: RefresherDemo(),
    );
  }
}

class RefresherDemo extends StatefulWidget {
  @override
  _RefresherDemoState createState() => _RefresherDemoState();
}

class _RefresherDemoState extends State<RefresherDemo> with SingleTickerProviderStateMixin {
  String _title = "Refresher Demo";
  List<Widget> _children = [];

  double percentage = 0.0;

  @override
  void initState() {
    // TODO: implement initState
    super.initState();
    _children = List.generate(
      Random().nextInt(100),
          (index) => Text("Here is $index"),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text(_title), actions: [
        IconButton(
          icon: Icon(Icons.navigate_next),
          onPressed: () {
            Navigator.push(context,
                MaterialPageRoute(builder: (BuildContext context) => InfiniteRefresherDemo()));
          },
        )
      ]),
      body: Refresher(
          loadingSize: 50.0,
          margin: EdgeInsets.all(8.0),
          onRefresh: () async {
            int counter = 0;
            Timer.periodic(Duration(milliseconds: 1000), (t) {
              counter++;
              if (this.mounted)
                setState(() {
                  _title = "Fetching${List.filled(counter, ".").join()}";
                  if (counter == 5) t.cancel();
                });
            });
            await Future.delayed(new Duration(milliseconds: 6000));
            if (this.mounted)
              setState(() {
                _title = "Refresher Demo";
                _children = List.generate(
                  Random().nextInt(100),
                  (index) => Text("Here is $index"),
                );
              });
          },
          child: Column(children: _children)),
    );
  }
}
